snapcraft/file_utils.py

```diff
@@ -110,6 +110,12 @@ def link_or_copy(source, destination, follow_symlinks=False):
         # symlinks.
         os.link(source_path, destination, follow_symlinks=False)
     except OSError:
+        try:
+            # If os.link raised an I/O error, it may have left a file behind.
+            if not os.path.isdir(destination):
+                os.unlink(destination)
+        except FileNotFoundError:
```

Comment:
I have no strong opinion about contextlib. I never use it, but it's not because I don't like it, it's because I never remember it exists.